src/frequenz/sdk/actor/power_distributing/battery_pool.py

```diff
@@ -0,0 +1,122 @@
+from typing import Set
+
+from ..._data_ingestion.battery_utils import BatteryUtils
+from .result import PartialFailure, Result, Success
+
+_logger = logging.getLogger(__name__)
+
+
+class BatteryPool:
+    """Holds bool of batteries and returns data from them."""
+
+    def __init__(
+        self,
+        battery_ids: Set[int],
+        max_data_age_sec: float,
+        max_blocking_duration_sec: float,
+    ) -> None:
+        """Create partially initialized object instance.
+
+        Note:
+            Please call `run` method to fully initialize BatteryPool. Otherwise
+            it is not possible to use BatteryPool.
+
+        Args:
+            battery_ids: set of batteries ids that should be stored in pool.
+            max_data_age_sec: If component stopped sending data, then
+                this is the maximum time when its last message should be considered as
+                valid. After that time, component won't be used until it starts sending
+                data.
+            max_blocking_duration_sec: This value tell what should be the maximum
+                timeout used for blocking failing component.
+        """
+        self._batteries = {
+            id: BatteryUtils(id, max_data_age_sec, max_blocking_duration_sec)
+            for id in battery_ids
+        }
+        self._run_method_called: bool = False
+
+    async def init(self) -> None:
+        """Run battery pool."""
+        tasks = [
+            asyncio.create_task(battery.init()) for battery in self._batteries.values()
+        ]
+        await asyncio.wait(tasks, return_when=ALL_COMPLETED)
+        self._run_method_called = True
+
+    def get_working_batteries(self, battery_ids: Set[int]) -> Set[int]:
+        """Get subset of battery_ids with working batteries.
+
+        Args:
+            battery_ids: batteries ids
+
+        Raises:
+            RuntimeError: If `run` method was not called at the beginning to
+                initialized object.
+            KeyError: If any battery in the given batteries is not in the pool.
+
+        Returns:
+            Subset of given batteries with working batteries.
+        """
+        if not self._run_method_called:
+            raise RuntimeError(
+                "`run` method not called or not awaited. Run it before first use"
+            )
+
+        ready: Set[int] = set()
+        blocked: Set[int] = set()
+        for bat_id in battery_ids:
+            if bat_id not in battery_ids:
+                ids = str(self._batteries.keys())
+                raise KeyError(f"No battery {bat_id} in pool. All batteries: {ids}")
+            if self._batteries[bat_id].is_ready():
+                ready.add(bat_id)
+            if self._batteries[bat_id].is_blocked():
+                blocked.add(bat_id)
+
+        working = ready - blocked
+
+        if len(working) > 0:
+            return working
+
+        _logger.warning("All batteries in %s are not working. Use blocked batteries.")
```

Comment:
The part `Use blocked batteries.` sounds like an instruction, and can be confusing. What is the intention here?

Reply:
Maybe something like "There are no working batteries in %s.  Falling back to using blocked batteries."?

Reply:
Done